File: bench/android/vosk/app/src/main/java/com/aidedx/voskbench/BenchActivity.java
package com.aidedx.voskbench;

import android.app.Activity;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.widget.Button;
import android.widget.TextView;

import org.vosk.LibVosk;
import org.vosk.LogLevel;
import org.vosk.Model;
import org.vosk.Recognizer;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Not a product UI (issue #120's non-goal) - a bare launcher activity that runs Vosk
 * file-based recognition over every WAV pushed under
 * getExternalFilesDir(null)/audio/&lt;speaker&gt;/&lt;id&gt;.wav against a model pushed to
 * getExternalFilesDir(null)/&lt;modelDir&gt;, and writes a results JSON in the same shape
 * scripts/asr-transcribe.mjs already produces (modelId/dtype/withPrompt/loadS/records[]) so
 * it drops straight into scripts/e2e-audio-intents.ts and scripts/asr-score-slots.mjs
 * unmodified.
 *
 * modelDir/outName/modelId come from Intent string extras (adb shell am start -e ...) so the
 * same APK benchmarks the EN and PL models without a rebuild.
 */
public class BenchActivity extends Activity {

    private TextView logText;
    private final StringBuilder log = new StringBuilder();

    @Override
    public void onCreate(Bundle state) {
        super.onCreate(state);
        setContentView(R.layout.main);
        logText = findViewById(R.id.log_text);
        logText.setMovementMethod(new ScrollingMovementMethod());
        Button runButton = findViewById(R.id.run_button);
        runButton.setOnClickListener(v -> new Thread(this::runBenchmark).start());
        // DEBUG (not INFO) so any "word not in vocabulary" grammar-compilation diagnostics
        // surface in logcat - issue #122's open question about OOV words in grammar mode.
        LibVosk.setLogLevel(LogLevel.DEBUG);
        appendLog("Ready. Tap \"Run benchmark\".");
        if (getIntent().getBooleanExtra("autorun", false)) {
            new Thread(this::runBenchmark).start();
        }
    }

    private void appendLog(String line) {
        log.append(line).append('\n');
        runOnUiThread(() -> logText.setText(log.toString()));
    }

    private void runBenchmark() {
        String modelDir = getIntent().getStringExtra("model_dir");
        if (modelDir == null) modelDir = "model-en";
        String outName = getIntent().getStringExtra("out_name");
        if (outName == null) outName = "results.json";
        String modelId = getIntent().getStringExtra("model_id");
        if (modelId == null) modelId = "vosk-" + modelDir;
        // Issue #122's open question: does restricting the decoder to a closed domain
        // vocabulary (Recognizer's grammar mode) recover jargon a free-form small model
        // mangles? grammarFile is a JSON word-list file already pushed into filesDir (same
        // way model/audio are), not an intent extra, since JSON needs no shell-escaping this way.
        String grammarFile = getIntent().getStringExtra("grammar_file");

        File base = getFilesDir();
        File modelPath = new File(base, modelDir);
        File audioBase = new File(base, "audio");
        File outFile = new File(base, outName);

        String grammarJson = null;
        int grammarWords = 0;
        if (grammarFile != null) {
            try {
                grammarJson = new String(java.nio.file.Files.readAllBytes(new File(base, grammarFile).toPath()));
                for (int i = 0; i < grammarJson.length(); i++) {
                    if (grammarJson.charAt(i) == ',') grammarWords++;
                }
                grammarWords++;
                appendLog("grammar: " + grammarFile + " (" + grammarWords + " words)");
            } catch (Exception e) {
                appendLog("FAILED to read grammar file " + grammarFile + ": " + e);
                return;
            }
        } else {
            appendLog("no grammar (free-form decoding)");
        }

        appendLog("model: " + modelPath);
        appendLog("audio: " + audioBase);

        long loadStart = System.nanoTime();
        Model model;
        try {
            model = new Model(modelPath.getAbsolutePath());
        } catch (Exception e) {
            java.io.StringWriter sw = new java.io.StringWriter();
            e.printStackTrace(new java.io.PrintWriter(sw));
            appendLog("FAILED to load model: " + sw);
            return;
        }
        double loadS = (System.nanoTime() - loadStart) / 1e9;
        appendLog(String.format("loaded in %.1fs", loadS));

        List<String> records = new ArrayList<>();
        File[] speakerDirs = audioBase.listFiles(File::isDirectory);
        if (speakerDirs == null) speakerDirs = new File[0];
        Arrays.sort(speakerDirs);

        for (File speakerDir : speakerDirs) {
            String speaker = speakerDir.getName();
            File[] wavs = speakerDir.listFiles((d, name) -> name.endsWith(".wav"));
            if (wavs == null) continue;
            Arrays.sort(wavs);
            for (File wav : wavs) {
                String id = wav.getName().substring(0, wav.getName().length() - 4);
                long t0 = System.nanoTime();
                String raw = "";
                String error = null;
                try {
                    raw = recognizeFile(model, wav, grammarJson);
                } catch (Exception e) {
                    error = String.valueOf(e.getMessage());
                }
                double secs = (System.nanoTime() - t0) / 1e9;
                records.add(jsonRecord(speaker, id, raw, secs, error));
                appendLog(String.format("%s/%s (%.2fs): %s", speaker, id,
                        secs, error != null ? "ERROR " + error : raw));
            }
        }
        model.close();

        String json = "{\n"
                + "  \"modelId\": " + jsonStr(modelId) + ",\n"
                + "  \"dtype\": \"device\",\n"
                + "  \"withPrompt\": false,\n"
                + "  \"grammar\": " + (grammarJson != null) + ",\n"
                + "  \"grammarWords\": " + grammarWords + ",\n"
                + "  \"loadS\": " + loadS + ",\n"
                + "  \"records\": [\n"
                + String.join(",\n", records)
                + "\n  ]\n"
                + "}\n";
        try (FileWriter w = new FileWriter(outFile)) {
            w.write(json);
        } catch (Exception e) {
            appendLog("FAILED to write results: " + e);
            return;
        }
        appendLog("wrote " + outFile + " (" + records.size() + " records)");
    }

    /** Feeds raw PCM16LE mono 16kHz bytes to a fresh Recognizer, chunked the same way the
     * upstream vosk-android demo does. When grammarJson is non-null, uses Recognizer's
     * grammar-mode constructor to restrict decoding to that closed word list instead of
     * free-form decoding. */
    private String recognizeFile(Model model, File wav, String grammarJson) throws Exception {
        Recognizer rec = grammarJson != null
                ? new Recognizer(model, 16000.f, grammarJson)
                : new Recognizer(model, 16000.f);
        byte[] bytes;
        try (InputStream in = new FileInputStream(wav)) {
            bytes = in.readAllBytes();
        }
        int dataOffset = findDataChunkOffset(bytes);
        byte[] buf = new byte[4096];
        for (int pos = dataOffset; pos < bytes.length; pos += buf.length) {
            int n = Math.min(buf.length, bytes.length - pos);
            System.arraycopy(bytes, pos, buf, 0, n);
            rec.acceptWaveForm(buf, n);
        }
        String finalJson = rec.getFinalResult();
        rec.close();
        return extractText(finalJson);
    }

    /** Locates the "data" chunk by scanning RIFF sub-chunks instead of assuming a fixed 44-byte
     * header - ffmpeg's pcm_s16le output embeds a LIST/INFO chunk pushing the real payload past
     * byte 44 for these resampled clips (same fix already used in the wav2vec2/whisper.cpp
     * benches, docs/android-asr-runtime-bench.md S4.1). */
    private static int findDataChunkOffset(byte[] bytes) {
        ByteBuffer header = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
        int pos = 12;
        while (pos + 8 <= bytes.length) {
            int id = header.getInt(pos);
            int size = header.getInt(pos + 4);
            if (id == 0x61746164) { // "data" little-endian
                return pos + 8;
            }
            pos += 8 + size + (size % 2);
        }
        throw new IllegalStateException("no \"data\" chunk found in " + bytes.length + " byte WAV");
    }

    /** Vosk returns {"text": "..."} (or {"text": ""} on silence) - avoid a JSON library
     * dependency for one field. */
    private static String extractText(String resultJson) {
        int i = resultJson.indexOf("\"text\"");
        if (i < 0) return "";
        int colon = resultJson.indexOf(':', i);
        int q1 = resultJson.indexOf('"', colon + 1);
        int q2 = resultJson.indexOf('"', q1 + 1);
        if (q1 < 0 || q2 < 0) return "";
        return resultJson.substring(q1 + 1, q2);
    }

    private static String jsonStr(String s) {
        return "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }

    private static String jsonRecord(String speaker, String id, String raw, double secs, String error) {
        return "    {\"speaker\": " + jsonStr(speaker) + ", \"id\": " + jsonStr(id)
                + ", \"raw\": " + jsonStr(raw) + ", \"secs\": " + secs
                + ", \"error\": " + (error == null ? "null" : jsonStr(error)) + "}";
    }
}

File: bench/android/sherpa-onnx/app/src/main/java/com/aidedx/sherpabench/BenchActivity.kt
package com.aidedx.sherpabench

import android.app.Activity
import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.widget.Button
import android.widget.TextView
import com.k2fsa.sherpa.onnx.FeatureConfig
import com.k2fsa.sherpa.onnx.OfflineModelConfig
import com.k2fsa.sherpa.onnx.OfflineRecognizer
import com.k2fsa.sherpa.onnx.OfflineRecognizerConfig
import com.k2fsa.sherpa.onnx.OfflineWhisperModelConfig
import com.k2fsa.sherpa.onnx.WaveReader
import java.io.File
import java.io.FileWriter
import java.io.PrintWriter
import java.io.StringWriter

/**
 * Not a product UI (issue #120's non-goal) - a bare launcher activity mirroring the Vosk bench
 * (bench/android/vosk): file-based recognition over every WAV pushed under
 * filesDir/audio/&lt;speaker&gt;/&lt;id&gt;.wav against a sherpa-onnx whisper-small int8 model
 * pushed to filesDir/&lt;modelDir&gt;, writing the same results JSON contract
 * scripts/asr-transcribe.mjs produces so it drops straight into
 * scripts/e2e-audio-intents.ts / scripts/asr-score-slots.mjs unmodified.
 *
 * Uses internal storage (filesDir) from the start, not external/adb-push-then-run-as - the
 * Vosk bench discovered adb push into Android/data/<pkg>/files isn't reliably readable by the
 * app itself under scoped storage (docs/android-asr-runtime-bench.md, S1). Same trick applies:
 * push to /data/local/tmp, then `adb shell run-as <pkg> cp -r ...` into filesDir.
 */
class BenchActivity : Activity() {

    private lateinit var logText: TextView
    private val log = StringBuilder()

    override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        setContentView(R.layout.main)
        logText = findViewById(R.id.log_text)
        logText.movementMethod = ScrollingMovementMethod()
        val runButton = findViewById<Button>(R.id.run_button)
        runButton.setOnClickListener { Thread { runBenchmark() }.start() }
        appendLog("Ready. Tap \"Run benchmark\".")
        if (intent.getBooleanExtra("autorun", false)) {
            Thread { runBenchmark() }.start()
        }
    }

    private fun appendLog(line: String) {
        log.append(line).append('\n')
        runOnUiThread { logText.text = log.toString() }
    }

    private fun runBenchmark() {
        val modelDir = intent.getStringExtra("model_dir") ?: "model-sherpa"
        val outName = intent.getStringExtra("out_name") ?: "results.json"
        val modelId = intent.getStringExtra("model_id") ?: "sherpa-onnx-$modelDir"
        val numThreads = intent.getIntExtra("num_threads", 4)

        val base = filesDir
        val modelPath = File(base, modelDir)
        val audioBase = File(base, "audio")
        val outFile = File(base, outName)

        appendLog("model: $modelPath")
        appendLog("audio: $audioBase")

        val config = OfflineRecognizerConfig(
            featConfig = FeatureConfig(sampleRate = 16000, featureDim = 80),
            modelConfig = OfflineModelConfig(
                whisper = OfflineWhisperModelConfig(
                    encoder = File(modelPath, "small-encoder.int8.onnx").absolutePath,
                    decoder = File(modelPath, "small-decoder.int8.onnx").absolutePath,
                    language = "en",
                    task = "transcribe",
                ),
                tokens = File(modelPath, "small-tokens.txt").absolutePath,
                modelType = "whisper",
                numThreads = numThreads,
                provider = "cpu",
            ),
        )

        val loadStart = System.nanoTime()
        val recognizer: OfflineRecognizer
        try {
            recognizer = OfflineRecognizer(config = config)
        } catch (e: Exception) {
            val sw = StringWriter()
            e.printStackTrace(PrintWriter(sw))
            appendLog("FAILED to load model: $sw")
            return
        }
        val loadS = (System.nanoTime() - loadStart) / 1e9
        appendLog("loaded in ${"%.1f".format(loadS)}s")

        val records = mutableListOf<String>()
        val speakerDirs = (audioBase.listFiles { f -> f.isDirectory } ?: arrayOf()).sortedBy { it.name }

        for (speakerDir in speakerDirs) {
            val speaker = speakerDir.name
            val wavs = (speakerDir.listFiles { _, name -> name.endsWith(".wav") } ?: arrayOf()).sorted()
            for (wav in wavs) {
                val id = wav.name.removeSuffix(".wav")
                val t0 = System.nanoTime()
                var raw = ""
                var error: String? = null
                try {
                    val waveData = WaveReader.readWave(wav.absolutePath)
                    recognizer.createStream().use { stream ->
                        stream.acceptWaveform(waveData.samples, waveData.sampleRate)
                        recognizer.decode(stream)
                        raw = recognizer.getResult(stream).text.trim()
                    }
                } catch (e: Exception) {
                    error = e.message
                }
                val secs = (System.nanoTime() - t0) / 1e9
                records.add(jsonRecord(speaker, id, raw, secs, error))
                appendLog("$speaker/$id (${"%.2f".format(secs)}s): ${error?.let { "ERROR $it" } ?: raw}")
            }
        }
        recognizer.release()

        val json = "{\n" +
            "  \"modelId\": ${jsonStr(modelId)},\n" +
            "  \"dtype\": \"device\",\n" +
            "  \"withPrompt\": false,\n" +
            "  \"loadS\": $loadS,\n" +
            "  \"records\": [\n" +
            records.joinToString(",\n") +
            "\n  ]\n" +
            "}\n"
        try {
            FileWriter(outFile).use { it.write(json) }
        } catch (e: Exception) {
            appendLog("FAILED to write results: $e")
            return
        }
        appendLog("wrote $outFile (${records.size} records)")
    }

    private fun jsonStr(s: String): String =
        "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\""

    private fun jsonRecord(speaker: String, id: String, raw: String, secs: Double, error: String?): String =
        "    {\"speaker\": ${jsonStr(speaker)}, \"id\": ${jsonStr(id)}, " +
            "\"raw\": ${jsonStr(raw)}, \"secs\": $secs, " +
            "\"error\": ${error?.let { jsonStr(it) } ?: "null"}}"
}
